src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -1278,7 +1308,15 @@ static BuildRequest CreateRealBuildRequest(BuildSubmission submission, int confi
                 _projectCacheService == null &&
                 _buildParameters.ProjectCacheDescriptor == null)
             {
-                _projectCacheServiceInstantiatedByVSWorkaround = true;
+                lock (_syncLock)
+                {
+                    if (_projectCacheServiceInstantiatedByVSWorkaround)
+                    {
```

Comment:
Setting something that's already true to true doesn't seem bad. I'd skip this lock and the if/return.